File: elab4-publication-backend/src/main/java/elaborate/publication/resources/AboutResource.java
package elaborate.publication.resources;

/*
 * #%L
 * elab4-publication-backend
 * =======
 * Copyright (C) 2013 - 2019 Huygens ING
 * =======
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as
 * published by the Free Software Foundation, either version 3 of the
 * License, or (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public
 * License along with this program.  If not, see
 * <http://www.gnu.org/licenses/gpl-3.0.html>.
 * #L%
 */

import java.io.IOException;
import java.util.Map;
import java.util.PropertyResourceBundle;

import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.Produces;

import com.google.common.collect.Maps;

import nl.knaw.huygens.jaxrstools.resources.UTF8MediaType;

@Path("about")
public class AboutResource {
  private static PropertyResourceBundle propertyResourceBundle;

  @GET
  //  @APIDesc("Get version info")
  @Produces(UTF8MediaType.APPLICATION_JSON)
  public Object getVersion() {
    Map<String, String> data = Maps.newHashMap();
    for (String field : new String[] { "commitId", "buildDate", "version", "scmBranch", "publishdate" }) {
      data.put("commitId", getProperty(field));
    }
    return data;
  }

  private static synchronized String getProperty(String key) {
    if (propertyResourceBundle == null) {
      try {
        propertyResourceBundle = new PropertyResourceBundle(Thread.currentThread().getContextClassLoader().getResourceAsStream("about.properties"));
      } catch (IOException e) {
        e.printStackTrace();
      }
    }
    return propertyResourceBundle.getString(key);
  }

}

File: elab4-publication-backend-app/src/main/kotlin/nl/knaw/huygens/elaborate/publication/resources/AboutResource.kt
package nl.knaw.huygens.elaborate.publication.resources

import java.io.File
import java.util.*
import javax.servlet.ServletContext
import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.Produces
import javax.ws.rs.core.Context
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.UriInfo
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import nl.knaw.huygens.elaborate.publication.AppConfig

@Path("/about")
class AboutResource(private val config: AppConfig) {
    val log: Logger = LoggerFactory.getLogger(AboutResource::class.java)

    private val stream = File("${config.publicationDir}/about.properties").inputStream()

    private val propertyResourceBundle: PropertyResourceBundle by lazy {
        PropertyResourceBundle(stream)
    }

    private val properties = listOf("commitId", "buildDate", "version", "scmBranch", "publishdate")

    @Produces(MediaType.APPLICATION_JSON)
    @GET
    fun get(
        @Context context: ServletContext,
        @Context uriInfo: UriInfo
    ): Any {
        val data: MutableMap<String, String> = mutableMapOf()
        val propertyKeys = propertyResourceBundle.keys.toList()
        for (field: String in properties) {
            if (propertyKeys.contains(field)) {
                data[field] = getProperty(field)
            }
        }
        data["serverInfo"] = context.serverInfo
        data["contextPath"] = context.contextPath
        data["baseUri"] = uriInfo.baseUri.toString()
        data["projectName"] = config.projectName
        data["publicationDir"] = config.publicationDir
        data["PROXY_URL"] = System.getenv("PROXY_URL") ?: "not defined"
        return data
    }

    @Synchronized
    private fun getProperty(key: String): String =
        propertyResourceBundle.getString(key)
}